### PR TITLE
GG-39685 Add ability to compress snapshot before encrypting them

### PR DESCRIPTION
**Pull request by ibessonov**
https://ggsystems.atlassian.net/browse/GG-39685

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java [hunk 1]
```diff
@@ -263,7 +263,10 @@ public enum IgniteFeatures {
     TASK_EVT_ATTRIBUTE_SUPPORT(69),
 
     /** Allows to node check connections with special heartbeat message to detect network issues rather quickly. */
-    TCP_COMMUNICATION_HEARTBEATS(70);
+    TCP_COMMUNICATION_HEARTBEATS(70),
+
+    /** Optimization of compression for encrypted snapshots. */
+    OPTIMIZED_COMPRESSED_ENCRYPTED_SNAPSHOTS(71);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java:269]: Forgot to add to 
https://ggsystems.atlassian.net/wiki/spaces/GG/pages/1192198276/Community+edition+features+list

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java [hunk 2]
```diff
@@ -2128,9 +2131,11 @@ public static boolean isPersistenceEnabled(DataStorageConfiguration cfg) {
      * @return Page size without encryption overhead.
      */
     public static int encryptedPageSize(int pageSize, EncryptionSpi encSpi) {
-        // If encryption is enabled, a space of one encryption block is reserved to store CRC and encryption key ID.
+        // If encryption is enabled, a space of one encryption block is reserved to store CRC, encryption key ID
+        // and position for EncryptedOutputStream.
         // If encryption is disabled, NoopEncryptionSPI with a zero encryption block size is used.
-        assert encSpi.blockSize() >= /* CRC */ 4 + /* Key ID */ 1 || encSpi.blockSize() == 0;
+        assert encSpi.blockSize() >= ENCRYPTION_CRC_AND_METADATA_SIZE + Integer.BYTES
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java:2137]:
```suggestion
        assert encSpi.blockSize() >= ENCRYPTION_CRC_AND_METADATA_SIZE + Integer.BYTES : encSpi.blockSize();
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java [hunk 3]
```diff
@@ -118,25 +118,36 @@ public void decrypt(ByteBuffer encrypted, ByteBuffer destBuf, GroupKey grpKey) t
      * @param res Destination buffer.
      */
     private void storeCRC(ByteBuffer res) {
-        int crc = FastCrc.calcCrc(res, encryptedDataSize());
+        int encryptedDataSize = encryptedDataSize();
+        int crc = FastCrc.calcCrc(res, encryptedDataSize);
+
+        assert res.position() == encryptedDataSize;
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java:124]:
```suggestion
        assert res.position() == encryptedDataSize : res.position();
```

**Review by tkalkirill — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java:269]: Added all missing features to the document

**Pull request merged**